benchmarks/newton_solver_benchmark_set/spiegelman_et_al_2016/drucker_prager_compositions.cc

```diff
@@ -0,0 +1,711 @@
+          for (unsigned int c=0; c < volume_fractions.size(); ++c)
+            thermal_conductivities += volume_fractions[c] * thermal_diffusivity[c] * heat_capacity[c] * densities[c];
+
+          // calculate effective viscosity
+          if (in.strain_rate.size())
+            {
+              // This function calculates viscosities assuming that all the compositional fields
+              // experience the same strain rate (isostrain). Since there is only one process in
+              // this material model (a general powerlaw) we do not need to worry about how to
+              // distribute the strain-rate and stress over the processes.
+              std::vector<double> composition_viscosities(volume_fractions.size());
+              std::vector<SymmetricTensor<2,dim> > composition_viscosities_derivatives(volume_fractions.size());
+              std::vector<double> composition_dviscosities_dpressure(volume_fractions.size());
+
+              const SymmetricTensor<2,dim> deviator_strain_rate = use_deviator_of_strain_rate ? deviator(in.strain_rate[i]) : in.strain_rate[i];
+
+              for (unsigned int c=0; c < volume_fractions.size(); ++c)
+                {
+                  // If strain rate is zero (like during the first time step) set it to some very small number
+                  // to prevent a division-by-zero, and a floating point exception.
+                  // Otherwise, calculate the square-root of the norm of the second invariant of the deviatoric-
+                  // strain rate (often simplified as epsilondot_ii)
+
+                  const double edot_ii = compute_second_invariant(deviator_strain_rate, min_strain_rate[c]);// std::max(edot_ii_strict, min_strain_rate[c]*min_strain_rate[c]);
+
+                  // Find effective viscosities for each of the individual phases
+                  // Viscosities should have same number of entries as compositional fields
+
+                  // Regularized Drucker Prager viscosity (see Spiegelman et al, 2016)
+                  composition_viscosities[c] = compute_viscosity(edot_ii,pressure,c,prefactor[c],true,min_visc[c],max_visc[c]);
+
+                  Assert(dealii::numbers::is_finite(composition_viscosities[c]),ExcMessage ("Error: Viscosity is not finite."));
+
+                  if (derivatives != NULL)
+                    {
+                      if (use_analytical_derivative)
+                        {
+                          //analytic
+                          if (c == 0  && composition_viscosities[c] <= max_visc[c] && composition_viscosities[c] >= min_visc[c])// && composition_viscosities[c] < max_visc[c] && composition_viscosities[c] > min_visc[c])
+                            {
+                              const double drucker_prager_viscosity = compute_viscosity(edot_ii,pressure,c,prefactor[c],false,min_visc[c],max_visc[c]);
+                              const double regulaization_adjustment = (ref_visc * ref_visc)
```

Comment:
We would also need to add this regularization in to make the benchmark work in `ViscoPlastic`.